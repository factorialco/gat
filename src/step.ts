export interface BaseStep {
  id?: string;
  name?: string;
  env?: Record<string, string>;
  ifExpression?: string;
  workingDirectory?: string;
}

export type Step = RunStep | UseStep;

export interface RunStep extends BaseStep {
  run: string;
}

export interface UseStep extends BaseStep {
  uses: string;
  with?: Record<string, string | number | boolean>;
}
