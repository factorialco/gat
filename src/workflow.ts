import { dump } from "js-yaml";
import kebabCase from "lodash/kebabCase";
import fs from "fs";
import path from "path";
import { promisify } from "util";
import axios from "axios";

import {
  ConcurrencyGroup,
  Job,
  JobOptions,
  StringWithNoSpaces,
  UsesJobOptions,
} from "./job";
import type { Event, EventName, EventOptions } from "./event";
import { type Step, isUseStep } from "./step";

const writeFilePromise = promisify(fs.writeFile);

interface DefaultOptions {
  workingDirectory: string;
}

interface EnvVar {
  name: string;
  value: string;
}

interface Tag {
  name: string;
  commit: {
    sha: string;
  };
}

export type RunnerDefinition =
  | string
  | { group: string; labels?: string[] }
  | ["self-hosted", string];

const chainAttackCache: Record<string, string> = {};

const supplyChainAttack = async (step: Step, enabled: boolean = false) => {
  if (!isUseStep(step)) return;

  if (!enabled) return step.uses;

  const uses = step.uses;

  if (!uses) return uses;

  if (chainAttackCache[uses]) return chainAttackCache[uses];

  const match = uses.match(/(?<repository>.*)@(?<version>.*)/);

  if (!match) return uses;

  const { repository, version } = match.groups as {
    repository: string;
    version: string;
  };

  const response = await axios.get(
    `https://api.github.com/repos/${repository}/tags`,
  );
  const tags = response.data as Tag[];

  const tag = tags.find((tag) => tag.name === version);

  if (!tag) return uses;

  const result = `${repository}@${tag.commit.sha}`;

  chainAttackCache[uses] = result;

  return result;
};

export class Workflow<
  JobStep extends Step = Step,
  Runner extends RunnerDefinition = RunnerDefinition,
  JobName = never,
> {
  events: Event[];
  jobs: Array<Job<JobStep, Runner, JobName>>;
  defaultOptions: DefaultOptions | null;
  env: EnvVar[];
  concurrencyGroup?: ConcurrencyGroup | null;

  constructor(public name: string) {
    this.events = [];
    this.jobs = [];
    this.defaultOptions = null;
    this.env = [];
  }

  on<T extends EventName>(name: T, options?: EventOptions<T>) {
    this.events = [...this.events, { name, options }];
    return this;
  }

  addDefaults(options: DefaultOptions) {
    this.defaultOptions = options;
    return this;
  }

  addJob<T extends string>(
    name: StringWithNoSpaces<T>,
    options: JobOptions<JobStep, Runner, JobName> | UsesJobOptions,
  ): Workflow<JobStep, Runner, JobName | T> {
    this.jobs = [...this.jobs, { name, options }];
    return this;
  }

  setEnv(name: string, value: string) {
    this.env = [...this.env, { name, value }];
    return this;
  }

  setConcurrencyGroup(concurrencyGroup: ConcurrencyGroup | null) {
    this.concurrencyGroup = concurrencyGroup;
    return this;
  }

  defaultRunner() {
    return "ubuntu-22.04";
  }

  async compile(filepath?: string) {
    const result = {
      name: this.name,
      on: Object.fromEntries(
        this.events.map(({ name, options }) => [
          name,
          options ? options : null,
        ]),
      ),
      concurrency: this.concurrencyGroup
        ? {
            group: this.concurrencyGroup.groupSuffix,
            "cancel-in-progress": this.concurrencyGroup.cancelPrevious,
          }
        : undefined,
      defaults: this.defaultOptions
        ? {
            run: {
              "working-directory": this.defaultOptions.workingDirectory,
            },
          }
        : undefined,
      env:
        this.env.length > 0
          ? Object.fromEntries(this.env.map(({ name, value }) => [name, value]))
          : undefined,
      jobs: Object.fromEntries(
        await Promise.all(
          this.jobs.map(async ({ name, options: jobOptions }) => {
            if ("uses" in jobOptions) {
              return [
                name,
                {
                  uses: jobOptions.uses,
                  with: jobOptions.with,
                  secrets: jobOptions.secrets,
                },
              ];
            }

            const {
              prettyName,
              permissions,
              ifExpression,
              runsOn,
              matrix,
              env,
              steps,
              dependsOn,
              services,
              timeout,
              concurrency,
              outputs,
              workingDirectory,
            } = jobOptions;

            return [
              name,
              {
                name: prettyName,
                permissions,
                if: ifExpression,
                "runs-on": runsOn ?? this.defaultRunner(),
                "timeout-minutes": timeout ?? 15,
                needs: dependsOn,
                services,
                concurrency: concurrency
                  ? {
                      group: `${kebabCase(this.name)}-${name}-${
                        concurrency.groupSuffix
                      }`,
                      "cancel-in-progress": concurrency.cancelPrevious,
                    }
                  : undefined,
                strategy: matrix
                  ? {
                      "fail-fast": false,
                      matrix:
                        typeof matrix === "string"
                          ? matrix
                          : {
                              ...Object.fromEntries(
                                matrix.elements.map(({ id, options }) => [
                                  id,
                                  options,
                                ]),
                              ),
                              include: matrix.extra,
                            },
                    }
                  : undefined,
                env,
                defaults: workingDirectory
                  ? {
                      run: {
                        "working-directory": workingDirectory,
                      },
                    }
                  : undefined,
                steps: await Promise.all(
                  steps.map(async (step) => {
                    const {
                      id,
                      name,
                      ifExpression,
                      workingDirectory,
                      continueOnError,
                      timeout,
                      ...options
                    } = step;
                    return {
                      id,
                      name,
                      if: ifExpression,
                      "continue-on-error": continueOnError,
                      "working-directory": workingDirectory,
                      "timeout-minutes": timeout,
                      ...options,
                      uses: await supplyChainAttack(step),
                    };
                  }),
                ),
                outputs,
              },
            ];
          }),
        ),
      ),
    };

    const compiled = `# Workflow automatically generated by gat\n# DO NOT CHANGE THIS FILE MANUALLY\n\n${dump(
      result,
      {
        noRefs: true,
        lineWidth: 200,
        noCompatMode: true,
      },
    )}`;

    if (!filepath) return compiled;

    return writeFilePromise(
      path.join(process.cwd(), ".github", "workflows", filepath),
      compiled,
    );
  }
}
