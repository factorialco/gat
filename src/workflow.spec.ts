import { describe, it, expect } from "vitest";
import { RunStep, UseStep } from "./step";
import { Workflow } from "./workflow";

describe("Workflow", () => {
  it("generates a simple workflow", () => {
    const workflow = new Workflow("Simple");
    workflow
      .on("pull_request", { types: ["opened"] })
      .addJob("job1", {
        steps: [{ name: "Do something", run: "exit 0" }],
      })
      .addJob("job2", {
        steps: [{ name: "Do something else", run: "exit 0" }],
        dependsOn: ["job1"],
      });

    expect(workflow.compile()).toMatchSnapshot();
  });

  it("allows multiple events", () => {
    const workflow = new Workflow("Multiple events");
    workflow
      .on("push", { branches: ["main"] })
      .on("pull_request", { types: ["opened"] })
      .addJob("job1", {
        steps: [{ name: "Do something", run: "exit 0" }],
      });
    expect(workflow.compile()).toMatchSnapshot();
  });

  it("allows declaring default options", () => {
    const workflow = new Workflow("Default options");
    workflow
      .on("push", { branches: ["main"] })
      .addDefaults({
        workingDirectory: "frontend",
      })
      .addJob("job1", {
        steps: [{ name: "Do something", run: "exit 0" }],
      });
    expect(workflow.compile()).toMatchSnapshot();
  });

  it("allows declaring environment variables", () => {
    const workflow = new Workflow("With Environment variables");
    workflow
      .on("push")
      .setEnv("NODE_VERSION", "16")
      .setEnv("ENABLED", "true")
      .addJob("job1", {
        steps: [
          {
            name: "Do something",
            run: "echo ${{ env.NODE_VERSION }} && echo ${{ env.ENABLED }}",
          },
        ],
      });
    expect(workflow.compile()).toMatchSnapshot();
  });

  it("allows using a concurrency group", () => {
    const workflow = new Workflow("Concurrency group");
    workflow.on("push").addJob("job1", {
      concurrency: {
        groupSuffix: "foo",
        cancelPrevious: true,
      },
      steps: [
        {
          run: "echo 0",
        },
      ],
    });
    expect(workflow.compile()).toMatchSnapshot();
  });

  it("allows using outputs", () => {
    const workflow = new Workflow("Using outputs");
    workflow.on("push").addJob("job1", {
      steps: [
        {
          id: "random-number",
          run: 'echo "::set-output name=random-number::$RANDOM"',
        },
      ],
      outputs: {
        "random-number": "${{ steps.random-number.outputs.random-number }}",
      },
    });
    expect(workflow.compile()).toMatchSnapshot();
  });

  it("allows conditional jobs", () => {
    const workflow = new Workflow("Conditional job");
    workflow.on("push").addJob("job1", {
      ifExpression: "${{ github.ref != 'refs/heads/main' }}",
      steps: [
        {
          run: "exit 0",
        },
      ],
    });
    expect(workflow.compile()).toMatchSnapshot();
  });

  it("allows a job matrix", () => {
    const workflow = new Workflow("Conditional job");
    workflow.on("push").addJob("job1", {
      matrix: {
        elements: [
          {
            id: "food",
            options: ["🍕", "🍔"],
          },
          {
            id: "topping",
            options: ["🍍"],
          },
        ],
        extra: [
          {
            food: "🍕",
            topping: "🍍",
            tasty: true,
          },
        ],
      },
      steps: [
        {
          run: "echo ${{ matrix.food }} with ${{ matrix.topping }}",
        },
      ],
    });
    expect(workflow.compile()).toMatchSnapshot();
  });

  it("allows uses steps", () => {
    const workflow = new Workflow("Uses steps");
    workflow
      .on("push")
      .setEnv("NODE_VERSION", "16")
      .setEnv("ENABLED", "true")
      .addJob("job1", {
        steps: [
          {
            uses: "actions/checkout@v3",
            with: {
              ref: "main",
            },
          },
        ],
      });
    expect(workflow.compile()).toMatchSnapshot();
  });

  it("allows custom types in a workflow", () => {
    interface MyUseStep extends UseStep {
      uses: "custom-action";
      with: { foo: string };
    }
    type CustomStep = RunStep | MyUseStep;
    type CustomRunner = "standard-runner";

    const workflow = new Workflow<CustomStep, CustomRunner>(
      "With custom types"
    );

    workflow.on("push").addJob("job1", {
      runsOn: "standard-runner",
      steps: [
        {
          run: "echo 'Do something'",
        },
        {
          uses: "custom-action",
          with: {
            foo: "bar",
          },
        },
      ],
    });

    expect(workflow.compile()).toMatchSnapshot();
  });

  it("support workflow dispatch event", () => {
    const workflow = new Workflow("Workflow dispatch");
    workflow
      .on("workflow_dispatch", {
        inputs: {
          foo: {
            description: "A foo input",
            required: true,
          },
          bar: {
            description: "A bar input",
            type: "choice",
            options: ["bar", "baz"],
          },
        },
      })
      .addJob("job1", {
        steps: [{ name: "Do something", run: "exit 0" }],
      });
    expect(workflow.compile()).toMatchSnapshot();
  });
});
