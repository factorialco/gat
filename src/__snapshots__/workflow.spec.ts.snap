// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Workflow > allows a job matrix 1`] = `
"# Workflow automatically generated by gat
# DO NOT CHANGE THIS FILE MANUALLY

name: Conditional job
on:
  push: null
jobs:
  job1:
    runs-on: ubuntu-22.04
    timeout-minutes: 15
    strategy:
      fail-fast: false
      matrix:
        food:
          - 🍕
          - 🍔
        topping:
          - 🍍
        include:
          - food: 🍕
            topping: 🍍
            tasty: true
    steps:
      - run: echo \${{ matrix.food }} with \${{ matrix.topping }}
"
`;

exports[`Workflow > allows concurrency groups at workflow level 1`] = `
"# Workflow automatically generated by gat
# DO NOT CHANGE THIS FILE MANUALLY

name: Concurrency at workflow level
on:
  push: null
concurrency:
  group: \${{ github.workflow }}-\${{ github.ref }}
  cancel-in-progress: true
jobs:
  job1:
    runs-on: ubuntu-22.04
    timeout-minutes: 15
    steps:
      - name: Do something
        run: exit 0
"
`;

exports[`Workflow > allows conditional jobs 1`] = `
"# Workflow automatically generated by gat
# DO NOT CHANGE THIS FILE MANUALLY

name: Conditional job
on:
  push: null
jobs:
  job1:
    if: \${{ github.ref != 'refs/heads/main' }}
    runs-on: ubuntu-22.04
    timeout-minutes: 15
    steps:
      - run: exit 0
"
`;

exports[`Workflow > allows creating jobs with concurrency set to null 1`] = `
"# Workflow automatically generated by gat
# DO NOT CHANGE THIS FILE MANUALLY

name: Without concurrency
on:
  push: null
jobs:
  job1:
    runs-on: ubuntu-22.04
    timeout-minutes: 15
    steps:
      - name: Do something
        run: exit 0
"
`;

exports[`Workflow > allows creating jobs with no steps, uses and secrets 1`] = `
"# Workflow automatically generated by gat
# DO NOT CHANGE THIS FILE MANUALLY

name: Uses job
on:
  push: null
jobs:
  job1:
    uses: example/example/.github/workflows/example1.yml@main
    with:
      foo: foo
    secrets: inherit
  job2:
    uses: example/example/.github/workflows/example2.yml@main
    with:
      bar: bar
    secrets:
      secret: super-secret
"
`;

exports[`Workflow > allows custom types in a workflow 1`] = `
"# Workflow automatically generated by gat
# DO NOT CHANGE THIS FILE MANUALLY

name: With custom types
on:
  push: null
jobs:
  job1:
    runs-on:
      - self-hosted
      - standard-runner
    timeout-minutes: 15
    steps:
      - run: echo 'Do something'
      - uses: custom-action
        with:
          foo: bar
"
`;

exports[`Workflow > allows declaring default options 1`] = `
"# Workflow automatically generated by gat
# DO NOT CHANGE THIS FILE MANUALLY

name: Default options
on:
  push:
    branches:
      - main
defaults:
  run:
    working-directory: frontend
jobs:
  job1:
    runs-on: ubuntu-22.04
    timeout-minutes: 15
    steps:
      - name: Do something
        run: exit 0
"
`;

exports[`Workflow > allows declaring environment variables 1`] = `
"# Workflow automatically generated by gat
# DO NOT CHANGE THIS FILE MANUALLY

name: With Environment variables
on:
  push: null
env:
  NODE_VERSION: '16'
  ENABLED: 'true'
jobs:
  job1:
    runs-on: ubuntu-22.04
    timeout-minutes: 15
    steps:
      - name: Do something
        run: echo \${{ env.NODE_VERSION }} && echo \${{ env.ENABLED }}
"
`;

exports[`Workflow > allows multiline strings 1`] = `
"# Workflow automatically generated by gat
# DO NOT CHANGE THIS FILE MANUALLY

name: Multiline strings
on:
  push: null
jobs:
  job1:
    runs-on: ubuntu-22.04
    timeout-minutes: 15
    steps:
      - name: Do something
        run: |-
          echo foo
          exit 0
"
`;

exports[`Workflow > allows multiple events 1`] = `
"# Workflow automatically generated by gat
# DO NOT CHANGE THIS FILE MANUALLY

name: Multiple events
on:
  push:
    branches:
      - main
  pull_request:
    types:
      - opened
jobs:
  job1:
    runs-on: ubuntu-22.04
    timeout-minutes: 15
    steps:
      - name: Do something
        run: exit 0
"
`;

exports[`Workflow > allows permissions into jobs 1`] = `
"# Workflow automatically generated by gat
# DO NOT CHANGE THIS FILE MANUALLY

name: Job with permissions
on:
  push: null
jobs:
  job1:
    permissions:
      contents: read
      pull-requests: write
    runs-on: ubuntu-22.04
    timeout-minutes: 15
    steps:
      - name: Do something
        run: exit 0
"
`;

exports[`Workflow > allows uses steps 1`] = `
"# Workflow automatically generated by gat
# DO NOT CHANGE THIS FILE MANUALLY

name: Uses steps
on:
  push: null
env:
  NODE_VERSION: '16'
  ENABLED: 'true'
jobs:
  job1:
    runs-on: ubuntu-22.04
    timeout-minutes: 15
    steps:
      - uses: actions/checkout@v3
        with:
          ref: main
"
`;

exports[`Workflow > allows using a concurrency group 1`] = `
"# Workflow automatically generated by gat
# DO NOT CHANGE THIS FILE MANUALLY

name: Concurrency group
on:
  push: null
jobs:
  job1:
    runs-on: ubuntu-22.04
    timeout-minutes: 15
    concurrency:
      group: concurrency-group-job1-foo
      cancel-in-progress: true
    steps:
      - run: echo 0
"
`;

exports[`Workflow > allows using outputs 1`] = `
"# Workflow automatically generated by gat
# DO NOT CHANGE THIS FILE MANUALLY

name: Using outputs
on:
  push: null
jobs:
  job1:
    runs-on: ubuntu-22.04
    timeout-minutes: 15
    steps:
      - id: random-number
        run: echo \\"::set-output name=random-number::$RANDOM\\"
    outputs:
      random-number: \${{ steps.random-number.outputs.random-number }}
"
`;

exports[`Workflow > generates a simple workflow 1`] = `
"# Workflow automatically generated by gat
# DO NOT CHANGE THIS FILE MANUALLY

name: Simple
on:
  pull_request:
    types:
      - opened
jobs:
  job1:
    runs-on: ubuntu-22.04
    timeout-minutes: 15
    steps:
      - name: Do something
        run: exit 0
  job2:
    runs-on: ubuntu-22.04
    timeout-minutes: 15
    needs:
      - job1
    steps:
      - name: Do something else
        run: exit 0
"
`;

exports[`Workflow > support workflow dispatch event 1`] = `
"# Workflow automatically generated by gat
# DO NOT CHANGE THIS FILE MANUALLY

name: Workflow dispatch
on:
  workflow_dispatch:
    inputs:
      foo:
        description: A foo input
        required: true
      bar:
        description: A bar input
        type: choice
        options:
          - bar
          - baz
jobs:
  job1:
    runs-on: ubuntu-22.04
    timeout-minutes: 15
    steps:
      - name: Do something
        run: exit 0
"
`;

exports[`Workflow > supports a pretty name for the job 1`] = `
"# Workflow automatically generated by gat
# DO NOT CHANGE THIS FILE MANUALLY

name: Job with pretty name
on:
  push: null
jobs:
  job1:
    name: My pretty name
    runs-on: ubuntu-22.04
    timeout-minutes: 15
    steps:
      - name: Do something
        run: exit 0
"
`;

exports[`Workflow > supports schedule event 1`] = `
"# Workflow automatically generated by gat
# DO NOT CHANGE THIS FILE MANUALLY

name: Schedule
on:
  schedule:
    - cron: 0 4 * * 1-5
jobs:
  job1:
    runs-on: ubuntu-22.04
    timeout-minutes: 15
    steps:
      - name: Do something
        run: exit 0
"
`;

exports[`Workflow > supports supply chain attack, with writing the lock file 1`] = `
"# Workflow automatically generated by gat
# DO NOT CHANGE THIS FILE MANUALLY

name: Supply chain attack
on:
  push: null
jobs:
  job1:
    runs-on: ubuntu-22.04
    timeout-minutes: 15
    steps:
      - name: Do something
        uses: tj-actions/changed-files@a284dc1814e3fd07f2e34267fc8f81227ed29fb8
"
`;

exports[`Workflow > supports supply chain attack, without writing the lock file 1`] = `
"# Workflow automatically generated by gat
# DO NOT CHANGE THIS FILE MANUALLY

name: Supply chain attack
on:
  push: null
jobs:
  job1:
    runs-on: ubuntu-22.04
    timeout-minutes: 15
    steps:
      - name: Do something
        uses: tj-actions/changed-files@youhavebeenhacked
"
`;
